controllers/imagehub/controllers/image_controller.go

```diff
@@ -61,159 +58,76 @@ type ImageReconciler struct {
-		Eventer:  r.Recorder,
-		Operator: r,
-		Gvk: schema.GroupVersionKind{
-			Group:   imagehubv1.GroupVersion.Group,
-			Version: imagehubv1.GroupVersion.Version,
-			Kind:    "Image",
-		},
-		FinalizerName: imagehubv1.ImgFinalizerName,
+	if err := r.Get(ctx, req.NamespacedName, img); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
-	img.APIVersion = ctr.Gvk.GroupVersion().String()
-	img.Kind = ctr.Gvk.Kind
 
-	return ctr.Run(ctx, req, img)
-}
-
-// Update .
-func (r *ImageReconciler) Update(ctx context.Context, req ctrl.Request, gvk schema.GroupVersionKind, obj client.Object) (ctrl.Result, error) {
-	r.Logger.V(1).Info("update reconcile controller image", "request", req)
-	img := &imagehubv1.Image{}
-	err := convert.JsonConvert(obj, img)
-	if err != nil {
-		r.Logger.V(2).Info("error in image json convert", "json", obj)
-		return ctrl.Result{Requeue: true}, err
+	if ok, err := r.finalizer.RemoveFinalizer(ctx, img, r.doFinalizer); ok {
+		return ctrl.Result{}, err
 	}
 
-	pipelines := []func(ctx context.Context, img *imagehubv1.Image){
-		r.syncRepo,
-	}
-	for _, fn := range pipelines {
-		fn(ctx, img)
+	if ok, err := r.finalizer.AddFinalizer(ctx, img); ok {
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return r.doReconcile(ctx, img)
 	}
-
-	return ctrl.Result{}, r.Client.Update(ctx, img)
+	return ctrl.Result{}, errors.New("reconcile error from Finalizer")
 }
 
-// Delete .
-func (r *ImageReconciler) Delete(ctx context.Context, req ctrl.Request, gvk schema.GroupVersionKind, obj client.Object) error {
-	r.Logger.V(1).Info("delete reconcile controller image", "request", req)
-	// get img from obj
-	img := &imagehubv1.Image{}
-	err := convert.JsonConvert(obj, img)
-	if err != nil {
-		r.Logger.V(2).Info("error in image json convert", "namespace/name", obj.GetNamespace()+"/"+obj.GetName())
-		return err
-	}
-
-	// todo try to delete image in hub.sealos.io registry
-
-	// get repo by img spec name
-	repo, err := r.db.getRepoByRepoName(ctx, img.Spec.Name.ToRepoName())
-	if err != nil {
-		r.Logger.V(2).Info("error in image getRepoByRepoName, not found repo by lable", "err:", err.Error())
-		return err
+func (r *ImageReconciler) doReconcile(ctx context.Context, obj client.Object) (ctrl.Result, error) {
+	r.Logger.V(1).Info("update reconcile controller image", "request", client.ObjectKeyFromObject(obj))
+	img, ok := obj.(*imagehubv1.Image)
+	if !ok {
+		return ctrl.Result{}, errors.New("obj convert Image is error")
 	}
-	// update repo spec
-	spec := imagehubv1.ReposiyorySpec{}
-	for _, t := range repo.Spec.Tags {
-		if t.Name != img.Spec.Name.GetTag() {
-			spec.Tags = append(spec.Tags, t)
+	// create repo and SetControllerReference.
+	repo := &imagehubv1.Repository{}
+	repo.Name = img.Spec.Name.ToRepoName().ToMetaName()
+	err := r.Get(ctx, client.ObjectKeyFromObject(repo), repo)
+	if err != nil && apierrors.IsNotFound(err) {
+		repo.Spec.Name = img.Spec.Name.ToRepoName()
+		if err := r.Create(ctx, repo); err != nil {
```

Comment:
If create image need to create repo, but create action failed, code continues to run?

Reply:
returned.